sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionProgressChecker.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction;
+
+import static java.math.MathContext.DECIMAL128;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.DONE;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.QUERY_CHANGE_STREAM;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.STOP;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.UPDATE_STATE;
+import static org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.PartitionMode.WAIT_FOR_CHILD_PARTITIONS;
+
+import com.google.cloud.Timestamp;
+import java.math.BigDecimal;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker.Progress;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class PartitionRestrictionProgressChecker {
+
+  private static final BigDecimal TOTAL_MODE_TRANSITIONS = BigDecimal.valueOf(3L);
+
+  /**
+   * Indicates how many mode transitions have been completed for the current mode. The transitions
+   * are as follows: * (1) QUERY_CHANGE_STREAM, (2) WAIT_FOR_CHILD_PARTITIONS, (3) FINISH_PARTITION,
+   * (4) WAIT_FOR_PARENT_PARTITIONS, (5) DELETE_PARTITION, (6) DONE.
+   *
+   * <p>This is used to calculate the units of work left, meaning that 1 transition = 1 unit of
+   * work.
+   */
+  private final Map<PartitionMode, BigDecimal> modeToTransitionsCompleted;
+
+  public PartitionRestrictionProgressChecker() {
+    modeToTransitionsCompleted = new HashMap<>();
+    modeToTransitionsCompleted.put(UPDATE_STATE, BigDecimal.valueOf(0L));
+    modeToTransitionsCompleted.put(QUERY_CHANGE_STREAM, BigDecimal.valueOf(1L));
+    modeToTransitionsCompleted.put(WAIT_FOR_CHILD_PARTITIONS, BigDecimal.valueOf(2L));
+    modeToTransitionsCompleted.put(DONE, BigDecimal.valueOf(3L));
+  }
+
+  public Progress getProgress(
```

Comment:
This is not up to date? Take a look at how we are calculation progress currently in https://github.com/apache/beam/blob/df67c817184acafc4adbc56a725eed4985053171/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/TimestampRangeTracker.java#L196-L235

Reply:
Hmmm, so TimestampRangeTracker sets the total work as either the number of seconds till now, or the number of seconds in the total restriction right? And the current work is the number of seconds in the most recently claimed position. And the remaining work is total work - current work.

I think my restriction tracker does the same, except it includes transition modes as extra units of work?

Reply:
Ah, you are doing this in the restriction tracker instead of in this class right? I think I missed that. Is there anyway we could have the full logic encapsulated in this class?

We should also include @hengfengli into the discussion, since he is the expert in this area.

Reply:
LGTM. The logic looks same to the current implementation. The only question is about why we need to add the number of transitions to the total work. It's a small issue as we are not required to have an absolute accurate progress here.

Reply:
Ah makes sense. Removed transitions from total work.